plutus-ledger/src/Ledger/Index.hs

```diff
@@ -151,21 +161,30 @@ lkpTxOut t = lookup t =<< ask
 validateTransaction :: ValidationMonad m
     => Slot.Slot
     -> Tx
-    -> m UtxoIndex
+    -> m (Maybe ValidationErrorInPhase, UtxoIndex)
 validateTransaction h t = do
-    _ <- checkSlotRange h t
-    _ <- checkValuePreserved t
-    _ <- checkPositiveValues t
-    _ <- checkFeeIsAda t
-
-    -- see note [Forging of Ada]
-    emptyUtxoSet <- reader (Map.null . getIndex)
-    unless emptyUtxoSet (checkForgingScripts t)
-    unless emptyUtxoSet (checkForgingAuthorised t)
-    unless emptyUtxoSet (checkTransactionFee t)
-
-    _ <- checkValidInputs t
-    insert t <$> ask
+    _ <- checkValidInputs (view inputsFees) t
+    (do
+        _ <- checkSlotRange h t
```

Comment:
Phase 2 is only for script validation, everything else (value preserved, positive value, witnesses for public key inputs, slot range) needs to be validated in phase 1.

Also I think we shoud check that `inputsFees`  and `inputs` don't overlap.